lib/parsers.js

```diff
@@ -53,123 +53,105 @@ module.exports.parseError = function parseError(err, kwargs, cb) {
   //   koa: req.url
-  //
   var originalUrl = req.originalUrl || req.url;
 
   // absolute url
-  var url = protocol + '://' + host + originalUrl;
+  var absoluteUrl = protocol + '://' + host + originalUrl;
 
-  // query string
-  //
+  // query string:
   //   node: req.url (raw)
-  //   express: req.query
-  //   koa: req.query
-  //
+  //   express, koa: req.query
   var query = req.query || urlParser.parse(originalUrl || '', true).query;
 
   // cookies:
-  //
-  //   node: req.headers.cookie
-  //   express: req.headers.cookie
-  //   koa: req.headers.cookie
-  //
+  //   node, express, koa: req.headers.cookie
   var cookies = cookie.parse(headers.cookie || '');
 
   // body data:
-  //
-  //   node: req.body
-  //   express: req.body
-  //   koa: req.body
-  //
+  //   node, express, koa: req.body
   var data = req.body;
   if (['GET', 'HEAD'].indexOf(method) === -1) {
     if (typeof data === 'undefined') {
       data = '<unavailable>';
     }
   }
 
-  if (data && {}.toString.call(data) !== '[object String]') {
+  if (data && typeof data !== 'string' && {}.toString.call(data) !== '[object String]') {
     // Make sure the request body is a string
     data = stringify(data);
   }
 
-  // client ip:
-  //
-  //   node: req.connection.remoteAddress
-  //   express: req.ip
-  //   koa: req.ip
-  //
-  var ip = req.ip || (req.connection || {}).remoteAddress;
-
   // http interface
   var http = {
     method: method,
     query_string: query,
     headers: headers,
     cookies: cookies,
     data: data,
-    url: url
+    url: absoluteUrl
   };
 
   // expose http interface
   kwargs.request = http;
 
-  // user
-  //
-  // typically found on req.user according to Express and Passport
-
-  var user = {};
-  if (!kwargs.user) {
-    if (req.user) {
-      // shallow copy is okay because we are only modifying top-level
-      // object (req.user)
-      for (var key in req.user) {
-        if ({}.hasOwnProperty.call(req.user, key)) {
-          user[key] = req.user[key];
+  // user: typically found on req.user in express/passport patterns
+  // five cases for parseUser value:
+  //   absent: grab only id, username, email from req.user
+  //   false: capture nothing
+  //   true: capture all keys from req.user
+  //   array: provided whitelisted keys to grab from req.user
+  //   function :: req -> user: custom parsing function
+  if (parseUser == null) parseUser = ['id', 'username', 'email'];
+  if (parseUser) {
+    var user = {};
+    if (typeof parseUser === 'function') {
+      user = parseUser(req);
+    } else if (req.user) {
+      if (parseUser === true) {
+        for (var key in req.user) {
+          if ({}.hasOwnProperty.call(req.user, key)) {
+            user[key] = req.user[key];
+          }
         }
+      } else {
+        parseUser.forEach(function (fieldName) {
+          if ({}.hasOwnProperty.call(req.user, fieldName)) {
+            user[fieldName] = req.user[fieldName];
+          }
+        });
```

Comment:
ehh...normally id agree but i did it that way at first and it felt like borderline overcommenting since each branch has a pretty straightforward condition check, so i made sure to describe the 5 cases together above instead